Add EIP-5625: NFT Metadata JSON Schema dStorage Extension (#5625)
ethereum/EIPs: EIPS/eip-5625.md

---
eip: 5625
title: NFT Metadata JSON Schema dStorage Extension
description: Add a dStorage property to non-fungible tokens (NFTs) metadata JSON schema to provide decentralized storage information of NFT assets
author: Gavin Fu (@gavfu)
discussions-to: https://ethereum-magicians.org/t/eip-5625-nft-metadata-json-schema-dstorage-extension/10754
status: Draft
type: Standards Track
category: ERC
created: 2022-09-08
requires: 721, 1155
---

## Abstract

> This standard is an extension to NFT metadata JSON schema defined by [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md). As a complement, it provides additional information about how the NFT asset is stored in a decentralized storage system, if applicable.

Reviewer: ```suggestion
This EIP extends the NFT metadata JSON schema defined in [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md), adding a `dStorage` key that provides information about how the NFT data is stored. 
```
I recommend you use this rephrasing of your current abstract. As written, the second sentence is grammatically incorrect.

As a refresher, let's review existing NFT metadata JSON schema standards. [EIP-721](./eip-721.md) defines a standard contract method `tokenURI` to return a given NFT's metadata JSON file, conforming to the *[EIP-721](./eip-721.md) Metadata JSON Schema*, which defines three properties: `name`, `description` and `image`.

Similarly, [EIP-1155](./eip-1155.md) also defines a standard contract method `uri` to return NFT metadata JSON files conforming to the *[EIP-1155](./eip-1155.md) Metadata JSON Schema*, which defines properties like `name`, `decimals`, `description`, `image`, `properties`, `localization`, etc.

Besides, as the world's largest NFT marketplace nowadays, OpenSea defines their own *Metadata Standards*, including a few more properties like `image_data`, `external_url`, `attributes`, `background_color`, `animation_url`, `youtube_url`, etc. This standard is de facto respected and followed by other NFT marketplaces like LooksRare.

> Apparently, none of these standards conveys storage information about the mime type asset that NFT token actually represents. This proposal is an effort to fill the missing part.

Reviewer: I spot a few grammar issues here. If you run this through something like Grammarly and fix the issues, I won't have to nitpick this and your EIP will be merged much quicker.

Author: Really appreciate your review. This section is checked via Grammarly and grammar issues are fixed accordingly. I'm not a native English speaker or writer, so if there are any more grammar issues to fix, or some place to rephrase, please don't hesitate to point out.  :)

                },
                "description": {
                    "type": "string",
                    "description": "A brief description of the dStorage platform"
                },
                "persistence_mechanism": {
                    "type": "string",
                    "description": "Persistence mechanism or incentive structure of the dStorage platform, like 'blockchain-based', 'contract-based', etc"
                },
                "challenge_mechanism": {
                    "type": "string",
                    "description": "Challenge mechanism of the dStorage platform, like Arweave's proof-of-access, etc"
                },
                "consensus": {
                    "type": "string",
                    "description": "Consensus mechanism of the dStorage platform, like PoW, PoS, etc"
                },
                "dstorage_note": {
                    "type": "string",
                    "description": "A note to prove the storage of the NFT asset on the dStorage platform, like a Filecoin deal id, a Crust place_storage_order transaction hash, etc"
                }
            }
        }
    }
}
```

> An example of an [EIP-721](./eip-721.md) NFT asset Metadata JSON file follows. The `dStorage` property contains information about how the NFT asset is stored on Crust Network, which can be verified on https://ipfs-scan.io?cid=QmVvigK7jfLyHcDjnNYr2mAReFW9FCFWf41UNB9tbK8rWw.
> 
> 
> ```json
> {
> 	"name": "LAND (13, -86)",
> 	"description": "A LAND is a digital piece of real estate in The Sandbox metaverse that players can buy to build experiences on top of",
> 	"image": "ipfs://QmVvigK7jfLyHcDjnNYr2mAReFW9FCFWf41UNB9tbK8rWw",
> 	"dStorage": {
> 		"platform": "Crust",
>     "description": "Crust Network, the Incentive Layer of IPFS",
>     "persistence_mechanism": "contract-based",
>     "challenge_mechanism": "mpow",
>     "consensus": "gpos",
>     "dstorage_note": "0x466ab180124de1718d30cd2e018e7fe013a07c4860674110ccd13e97eb31ae16"
> 	}
> }
> ```
> 

Reviewer: ```suggestion
```
This should either be in the reference implementation section or omitted. If you would prefer your EIP to be merged more quickly, I would recommend just omitting it.

## Rationale

### Choice between Interface and JSON Schema Extension

> A smart contract interface like `NFTdStorageMetadata` would require extra implementation, which is detrimental to the standard adoption, especially for NFT projects that already have their NFT smart contracts finalized and deployed. By contrast, an optional JSON schema extension is noninvasive, and more easily to be adopted.

Reviewer: ```suggestion
An extension of the EIP-721 or EIP-1155 contract interfaces would unnecessarily require additional code to implement, and would not be available for use by NFT projects that already have their NFT smart contracts finalized and deployed. An optional JSON schema extension is noninvasive, and more easily adopted.
```
I recommend you rephrase this as above

# Backwards Compatibility

> This EIP is backward compatible with [EIP-721](./eip-721.md)  and [EIP-1155](./eip-1155.md) .

Reviewer: ```suggestion
This EIP is backward compatible with [EIP-721](./eip-721.md)  and [EIP-1155](./eip-1155.md).
```
Remove trailing space.

## Security Considerations

> There are no security considerations related directly to the implementation of this standard.

Reviewer: ```suggestion
Needs discussion.
```
I wouldn't be able to say *how* exactly this could cause any security considerations (maybe a malicious contract says that the metadata is immutable, but it's not?), but a draft isn't considered stable enough to definitively say that there *are* ***no*** security considerations.
